ArjunaCore/arjuna/tests/classes/com/hp/mwtests/ts/arjuna/reaper/ReaperTestCase4.java

```diff
@@ -0,0 +1,93 @@
+package com.hp.mwtests.ts.arjuna.reaper;
+
+import com.arjuna.ats.arjuna.common.Uid;
+import com.arjuna.ats.arjuna.coordinator.Reapable;
+import com.arjuna.ats.arjuna.coordinator.TransactionReaper;
+import org.jboss.byteman.contrib.bmunit.BMScript;
+import org.jboss.byteman.contrib.bmunit.BMUnitConfig;
+import org.jboss.byteman.contrib.bmunit.BMUnitRunner;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.junit.Assert.fail;
+
+@RunWith(BMUnitRunner.class)
+@BMScript("reaper")
+@BMUnitConfig(debug = true)
+public class ReaperTestCase4 extends ReaperTestCaseControl {
+
+    private static int _timeout = 5;
+
+    private Reapable reapable1 = new SimpleReapable();
+    private Reapable reapable2 = new SimpleReapable();
+
+    @Test
+    public void test() throws InterruptedException {
+        enableRendezvous("controlledshutdown", false);
+        TransactionReaper transactionReaper = TransactionReaper.transactionReaper();
+        transactionReaper.insert(reapable1, _timeout);
+        // This second Reapable must not be removed as we should check if TransactionReaper waits for it
+        // even when waitForTransaction is false
+        transactionReaper.insert(reapable2, _timeout);
+
+        AtomicBoolean terminated = new AtomicBoolean(false);
+        new Thread(() -> {
+            // The shutdown is from another thread
+            transactionReaper.terminate(false);
+            synchronized (terminated) {
+                terminated.set(true);
+                terminated.notify();
+            }
+        }).start();
+        // This needs to be after the thread above has started so that the other thread can start a shutdown
+        transactionReaper.remove(reapable1);
+
+        synchronized (terminated) {
+            if (!terminated.get()) {
+                try {
+                    // Give some time for the terminate to finish
+                    long beforeWait = System.currentTimeMillis();
+                    terminated.wait();
+                    long afterWait = System.currentTimeMillis();
+                    long duration = afterWait - beforeWait;
+
+                    // The duration (_timeout * 1000) + 1000 is worked out by adding 1s to _timeout. This is needed
+                    // as ReaperThread needs to wait the first timeout and then process the other transactions.
+                    Assert.assertTrue(String.format("The timeout duration was %s!", duration), duration < ((_timeout * 1000) + 1000));
```

Comment:
As mentioned elsewhere, another thread is waking up the terminate from an unexpected place (https://github.com/jmfinelli/narayana/blob/7ee80804e9bd13364d57d22bc67d3499f74336cf/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java#L828) and even if I pause the TransactionReaperThread from doing the notify I can see that TransactionReaper is sleeping here  https://github.com/jmfinelli/narayana/blob/7ee80804e9bd13364d57d22bc67d3499f74336cf/ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/coordinator/ReaperThread.java#L63. For some reason, it is sleeping there for a very long time and is not waking up in a reasonable time to be able to call the check method but I presume even if it were waking up after some time that I could expect it would be at the next scheduled timeout and that could still be a long wait (and more than the 1000 millis that the shutdown would wait for today)?

Reply:
I hope I replied to this question with [this](https://github.com/tomjenkinson/narayana/pull/4/files#r1494772629) comment

Reply:
As things stand, this test takes (almost) as long as the timeout period of the Reapable for it to complete (if you print out the duration in some new code you will find it will be close to 5000 millis).

Given how things are with https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java#L735, the new current expectation is that the terminate (when `waitForTransactions` is `false`, as we have in this test) should be basically something close to 1000 millis even if _inShutdown were read from a cache by the thread removing a ReaperElement.

It may be that the delay is related to my not seeing the notify() not being fired in a timely manner that I mentioned in my thread in TransactionReaper.java? Maybe because the Reaper Thread is sleeping and so can't enter check() in a timely manner.

Reply:
>  the Reaper Thread is sleeping and so can't enter check() in a timely manner.

That should be the reason. I think this is a reflection of the [original design](https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java#L779) of the TransactionReaper

Reply:
I don't think that is necessarily the way to read that comment, I think it basically means let the reaper work up as it was going to (at the timeout of the element) and not find the element (and then I presume go back to sleep) rather than recalculating the wake up time to be the next longest timeout pre-emptively.

I think the goal for `waitForTransactions` being `false` should be to ensure a timely wake up.

Reply:
> > I think the goal for waitForTransactions being false should be to ensure a timely wake up.
> 
> That's what happens, **usually**. OTOH, the backup logic defined in `JBTM-3830_jfinelli` will be triggered very very rarely thus it shouldn't be a problem waiting for the shortest timeout. Having said that, if we want to shut down immediately in any possible case we can set `nextDynamicCheckTime` to zero immediately after we call `setAllTimeoutsToZero()`.

I presume that the expected purpose of the changes here are essentially to verify that we can shut down immediately the case of a potential cached read of _inShutdown? I have tried setting `nextDynamicCheckTime` to zero immediately after we call `setAllTimeoutsToZero()` but that didn't seem to work for me: https://github.com/jmfinelli/narayana/compare/jmfinelli:JBTM-3830_jfinelli...tomjenkinson:JBTM-3830_jfinelli (the test change sets a `passing` "benchmark" duration to be 2000 millis in order to be comfortably over the wait of [1000](https://github.com/jbosstm/narayana/blob/d6004f4/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java#L735) that the main branch currently waits for).

Reply:
Forgot to mention `this.notify()`. I pushed another to shut down immediately

Reply:
Thank you - that was the missing part for getting this to work for me!!

I was able to trigger what looks like a blocking wait situation (I am wondering if holding it in the debugger passed the scheduled timeout might be the cause due to I think it uses absolute timeouts somehow?) but I don't say that is caused by this change (it would need some more investigation. If it is the case, though maybe the wait(1000) we have in main would cover that situation too?

Reply:
It might not be that, I have a stack dump from another run where a lock happened and it might be something in the Byteman rules with the rendezvous:

```
"main" #1 prio=5 os_prio=0 cpu=420.72ms elapsed=75.15s tid=0x00007faec002e5f0 nid=0x87d waiting for monitor entry  [0x00007faec8114000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at com.arjuna.ats.arjuna.coordinator.TransactionReaper.removeElementClient(TransactionReaper.java:807)
	- waiting to lock <0x000000062b8ef960> (a com.arjuna.ats.arjuna.coordinator.TransactionReaper)
	at com.arjuna.ats.arjuna.coordinator.TransactionReaper.remove(TransactionReaper.java:637)
	- locked <0x000000062b9094a0> (a com.arjuna.ats.internal.arjuna.coordinator.ReaperElement)
	at com.hp.mwtests.ts.arjuna.reaper.ReaperTestCase4.test(ReaperTestCase4.java:55)
	at com.hp.mwtests.ts.arjuna.reaper.ReaperTestCase4.testWithRendezvous(ReaperTestCase4.java:32)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@17.0.10/Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@17.0.10/NativeMethodAccessorImpl.java:77)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@17.0.10/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@17.0.10/Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.jboss.byteman.contrib.bmunit.BMUnitRunner$6.evaluate(BMUnitRunner.java:263)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.jboss.byteman.contrib.bmunit.BMUnitRunner$2.evaluate(BMUnitRunner.java:126)
	at org.jboss.byteman.contrib.bmunit.BMUnitRunner$1.evaluate(BMUnitRunner.java:97)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)

"Transaction Reaper" #20 daemon prio=5 os_prio=0 cpu=4.40ms elapsed=74.16s tid=0x00007faec04e2510 nid=0x8ba waiting for monitor entry  [0x00007fae90cf4000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at com.arjuna.ats.arjuna.coordinator.TransactionReaper.check(TransactionReaper.java:96)
	- waiting to lock <0x000000062b8ef960> (a com.arjuna.ats.arjuna.coordinator.TransactionReaper)
	at com.arjuna.ats.internal.arjuna.coordinator.ReaperThread.run(ReaperThread.java:79)

"Transaction Reaper Worker 0" #21 daemon prio=5 os_prio=0 cpu=0.57ms elapsed=74.16s tid=0x00007faec04e7640 nid=0x8bb in Object.wait()  [0x00007fae90bf4000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@17.0.10/Native Method)
	- waiting on <0x000000062b8f6528> (a java.util.LinkedList)
	at java.lang.Object.wait(java.base@17.0.10/Object.java:338)
	at com.arjuna.ats.arjuna.coordinator.TransactionReaper.waitForWork(TransactionReaper.java:344)
	- locked <0x000000062b8f6528> (a java.util.LinkedList)
	at com.arjuna.ats.internal.arjuna.coordinator.ReaperWorkerThread.run(ReaperWorkerThread.java:48)

"Thread-1" #22 prio=5 os_prio=0 cpu=3.46ms elapsed=61.85s tid=0x00007faec04e8820 nid=0x8bf in Object.wait()  [0x00007fae917fc000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@17.0.10/Native Method)
	- waiting on <0x000000062c3b9de0> (a org.jboss.byteman.synchronization.Rendezvous)
	at java.lang.Object.wait(java.base@17.0.10/Object.java:338)
	at org.jboss.byteman.synchronization.Rendezvous.rendezvous(Rendezvous.java:74)
	at org.jboss.byteman.rule.helper.Helper.rendezvous(Helper.java:972)
	- locked <0x000000062c3b9de0> (a org.jboss.byteman.synchronization.Rendezvous)
	at org.jboss.byteman.rule.helper.Helper.rendezvous(Helper.java:954)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@17.0.10/Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@17.0.10/NativeMethodAccessorImpl.java:77)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@17.0.10/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@17.0.10/Method.java:568)
	at org.jboss.byteman.rule.expression.MethodExpression.interpret(MethodExpression.java:366)
	at org.jboss.byteman.rule.Action.interpret(Action.java:144)
	at org.jboss.byteman.rule.helper.InterpretedHelper.fire(InterpretedHelper.java:170)
	at org.jboss.byteman.rule.helper.InterpretedHelper.execute0(InterpretedHelper.java:138)
	at org.jboss.byteman.rule.helper.InterpretedHelper.execute(InterpretedHelper.java:100)
	at org.jboss.byteman.rule.Rule.execute(Rule.java:820)
	at org.jboss.byteman.rule.Rule.execute(Rule.java:789)
	at com.arjuna.ats.arjuna.coordinator.TransactionReaper.shutdown(TransactionReaper.java:754)
	- locked <0x000000062b8ef960> (a com.arjuna.ats.arjuna.coordinator.TransactionReaper)
	at com.arjuna.ats.arjuna.coordinator.TransactionReaper.terminate(TransactionReaper.java:975)
	- locked <0x000000062bc194a0> (a java.lang.Class for com.arjuna.ats.arjuna.coordinator.TransactionReaper)
	at com.hp.mwtests.ts.arjuna.reaper.ReaperTestCase4.lambda$test$0(ReaperTestCase4.java:48)
	at com.hp.mwtests.ts.arjuna.reaper.ReaperTestCase4$$Lambda$89/0x00007fae330d2778.run(Unknown Source)
	at java.lang.Thread.run(java.base@17.0.10/Thread.java:840)
```

Reply:
Cool. I'm happy it worked out well :-)